cli/fathom_web/commands/train.py

```diff
@@ -46,18 +51,53 @@ def learn(learning_rate, iterations, x, y, confidence_threshold, num_prunes, val
                         previous_validation_loss = validation_loss
                         previous_model = model.state_dict()
                 writer.add_scalar('validation_loss', validation_loss, t)
-            accuracy, _, _ = accuracy_per_tag(y, y_pred, confidence_threshold, num_prunes)
+            accuracy, _, _ = accuracy_per_tag(y, y_pred, cutoff=0.5, num_prunes=num_prunes)
             writer.add_scalar('training_accuracy_per_tag', accuracy, t)
             optimizer.zero_grad()  # Zero the gradients.
             loss.backward()  # Compute gradients.
             optimizer.step()
     if stopped_early:
         print(f'Stopping early at iteration {t}, just before validation error rose.')
 
+    optimal_cutoff = find_optimal_cutoff(y, y_pred, num_prunes)
     # Horizontal axis is what confidence. Vertical is how many samples were that confidence.
     writer.add_histogram('confidence', confidences(model, x), t)
     writer.close()
-    return model
+    return model, optimal_cutoff
+
+
+def get_possible_cutoffs(y_pred):
+    """Using y_pred gets the sigmoid values, rounds the values and then gets the unique list.
+    This will reduce the number of cutoffs to be evaluated."""
+    with torch.no_grad():
+        flattened = y_pred.sigmoid().numpy().flatten()
+        cutoffs = np.sort(flattened)
+        if len(cutoffs) > 1:
+            cutoffs = np.unique([round((current + next) / 2, CUTOFF_DECIMAL_PLACES) for current, next in pairwise(cutoffs)])
+        return cutoffs
```

Comment:
Does the right thing happen if there's only 1 (unique) datapoint? I think it could go either way. Let's say every datapoint had a value of 0.5, and some were positives (T) and some negatives (F). accuracy_per_tag() calls it a positive if the datapoint ≥ the cutoff. So this set would work out in our favor:
`0.5T 0.5T 0.5T`

And this set would be all wrong:
`0.5F 0.5F 0.5F`

So really, we always need to consider at least 2 cutoffs: in the case of a single unique value, some iota to the left and some iota to the right.

I'll leave this part as an exercise to the reader, but do you think we always need to consider a possible cutoff position an iota to the left of the min value and to the right of the max? It might be possible to construct a pathological dataset that needs those in the case of multiple unique values, and, if it is useful there, it would save a special case.

Reply:
Determining the optimal cutoff using a single (or 2 or 3) possible cutoff isn't adequate.  I'll look into a method to expand the list of possible cutoffs.

WRT always needing to consider a possible cutoff to the left on the min or right of the max:  

The only time adding another possible threshold higher than the calculated max possible threshold would have an impact on accuracy is if the optimal threshold is the highest threshold from the possibles. e.g [0.3, 0.4, 0.5] Adding 0.55 to the possibles list does not add any benefit if 0.4 is the optimal value.  If 0.5 is the optimal value then adding 0.55 to the possibles list may result in an increase in accuracy.

However the value of 0.5 is the midpoint of the sigmoid values for the 2 highest y_pred values.  Assuming that y is labeled a positive for the highest prediction value, moving the cutoff more to the right will not have an impact unless it goes beyond the sigmoid value of the highest y_pred.  That will result in all negatives giving a worse accuracy that the midpoint cutoff.  Does this make sense?  am I over simplifying?

I tested using a variety of y_pred values from [-2:2) inc by 0.01.  I also used [-2:1) incr by 0.01 with [2] appended to provide a large gap in the values at the high end. I could not find a dataset that showed improvement by adding a possible cutoff to the right of the max possible cutoff.

Reply:
To solve the problem of not having an adequate number of possible cutoff values I want to propose:

If there are less than a predetermined number of values (say 5),  for each possible cutoff pull 10 values from a normal distribution, truncated to [0,1] with mu set to the existing possible cutoff value and sigma low (current testing uses 0.1)  These values are then run through the same rounding and uniqueness processing to reduce the list.  This gives additional possible cutoffs that are close to the data driven cutoffs.